### PR TITLE
Add banner configurable via module API

### PR DESCRIPTION
**Pull request by maheichyk**
Type: Enhancement
Related: https://github.com/matrix-org/matrix-react-sdk-module-api

This PR suggests to add possibility to configure a banner at the top of Element that is configurable via module API.

This could be used to show custom menu, widget toggle buttons or other information that makes sense in custom fork deployments.

There is no module API for that currently but could look like this:
```
export enum BannerLifecycle {
    Banner = "banner",
}

export type BannerOpts = {
    banner: JSX.Element | undefined;
};

export type BannerListener = (opts: BannerOpts) => void;
```

Module demo that shows simply button in the banner:
```
protected onBanner: BannerListener = (opts) => {
    opts.banner = <button>Custom Menu</button>;
};
```

And look the following:
Screenshot 2023-03-08 at 10 45 53

## Checklist

-   [ ] Tests written for new code (and old code if feasible)
-   [ ] Linter and other CI checks pass
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add banner configurable via module API ([\#10321](https://github.com/matrix-org/matrix-react-sdk/pull/10321)). Contributed by @maheichyk.

### CODE REVIEW

**Comment by maheichyk** [comment]
Would be great to hear some feedback on this proposal.

**Comment by maheichyk** [comment]
Could you please provide some feedback on this proposal?

File: src/components/structures/MatrixChat.tsx [hunk 1]
```diff
@@ -2135,6 +2137,17 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
             return null;
         }
 
+        const opts: BannerOpts = { banner: undefined };
+        ModuleRunner.instance.invoke(BannerLifecycle.Banner, opts);
+        if (opts.banner) {
+            view = (
+                <div className="mx_MatrixChat_wrapper">
```
Comment (gsouquet) [inline src/components/structures/MatrixChat.tsx:2144]: From an implementation point of view, I don't think we would need to add a wrapper as the banner could be slotted as a sibling of the `MatrixChat` and the parent section `id="matrixchat"` is a flex container that applies the same CSS that you declared for `mx_MatrixChat_wrapper`. 
I'd be advocating for a leaner HTML structure

**Review by gsouquet — CHANGES_REQUESTED** [review]
From a technical point of view, I believe this can be a bit leaner as the HTML structure already exists and could be re-used.

We would need a test validating that a banner can be injected in the page too

But more importantly this requires a product review

**Comment by daniellekirkwood** [comment]
👋  Product here... 

Regular users would not see this? This is just available for power users & people running a fork..?

**Comment by maheichyk** [comment]
> 👋 Product here...
> 
> Regular users would not see this? This is just available for power users & people running a fork..?

Hello :wave: yes, nothing is changed for the regular users. This is intended to be used in custom Element deployments. Currently we have to fork Element and apply custom changes. With this change in place, we could write a module that will add the banner useful in our deployment.

**Comment by t3chguy** [comment]
The product consideration here is one of bandwidth & maintenance, by adding this to the module API we have to ensure we don't break it and maintain it if someone else does

**Comment by daniellekirkwood** [comment]
Yes, comfortable with this. Thanks both!

**Review by daniellekirkwood — APPROVED** [review]

File: src/components/structures/MatrixChat.tsx (hunk 1 above)
Comment (maheichyk) [reply src/components/structures/MatrixChat.tsx]: Changed implementation to place the Banner as a sibling of the `mx_MatrixChat`. Had to update existing `mx_MatrixChat_wrapper` with `flex-direction: column` for the Banner to be shown on the top of Element. 

I have created a PR to the module API with a suggestion to add the Banner: https://github.com/matrix-org/matrix-react-sdk-module-api/pull/13

**Comment by weeman1337** [comment]
@maheichyk can you update the code here and fix the tests? I would then have a look at your PRs.

**Comment by maheichyk** [comment]
Another PR is created to element-web: https://github.com/vector-im/element-web/pull/25537. It is based on ideas from this comment: https://github.com/matrix-org/matrix-react-sdk/pull/10321/#discussion_r1132097998. Element web seems to be a better place for these changes, please have a look.

**Comment by germain-gg** [comment]
Superseeded by https://github.com/vector-im/element-web/pull/25537

**Pull request closed without merging**